File: solutions/cpp/Power_set_alternative.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cmath>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::ostream_iterator;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

void GeneratePowerSetHelper(const vector<int>& S, int m, vector<int>* subset);

// @include
void GeneratePowerSet(const vector<int>& S) {
  vector<int> subset;
  GeneratePowerSetHelper(S, 0, &subset);
}

void GeneratePowerSetHelper(const vector<int>& S, int m,
                            vector<int>* subset) {
  if (!subset->empty()) {
    // Print the subset.
    cout << subset->front();
    for (int i = 1; i < subset->size(); ++i) {
      cout << "," << (*subset)[i];
    }
  }
  cout << endl;

  for (int i = m; i < S.size(); ++i) {
    subset->emplace_back(S[i]);
    GeneratePowerSetHelper(S, i + 1, subset);
    subset->pop_back();
  }
}
// @exclude

int main(int argc, char* argv[]) {
  vector<int> S;
  if (argc >= 2) {
    for (int i = 1; i < argc; ++i) {
      S.emplace_back(atoi(argv[i]));
    }
  } else {
    default_random_engine gen((random_device())());
    uniform_int_distribution<int> dis(1, 10);
    S.resize(dis(gen));
    for (int i = 0; i < S.size(); ++i) {
      S[i] = i;
    }
  }
  GeneratePowerSet(S);
  return 0;
}

File: solutions/python/power_set_alternative.py
# Power_set_alternative.cpp 848813e190b1b85a8e75107fe8513c3be38ad1a9
import sys
import random
import itertools


# @include
def generate_power_set(S):
    subset = []
    generate_power_set_helper(S, 0, subset)

def generate_power_set_helper(S, m, subset):
    # Print the subset.
    print(*subset, sep=',')

    for i in range(m, len(S)):
        subset.append(S[i])
        generate_power_set_helper(S, i + 1, subset)
        subset.pop()
# @exclude


# Pythonic solution
def generate_power_set_pythonic(S):
    for n in range(len(S)+1):
        for i in itertools.combinations(S, n):
            print(*i, sep=',')


def main():
    if len(sys.argv) >= 2:
        S = [int(i) for i in sys.argv[1:]]
    else:
        S = list(range(random.randint(1, 10)))
    generate_power_set(S)
    generate_power_set_pythonic(S)


if __name__ == '__main__':
    main()
